### PR TITLE
Implement ZeroizeOnDrop for SHA 1..=2 and Blake2

### PR DESCRIPTION
**Pull request by kayabaNerve**
Relevant to #87.

Resets the hasher state after calling zeroize so it's still usable (1 and 2) indistinguishable from a newly constructed hasher.

This needs a impl Zeroize where Wrapped: Zeroize on CoreWrapper and associated in digest to be pleasantly effective. It's technically reachable now via wrappers' `decompose` functions. I would've done that with this PR yet the digest crate is in a distinct repo.

### CODE REVIEW

**Comment by newpavlov** [comment]
Do not implement the `Zeroize` trait for cryptographic types. Users may execute it and get an invalid state. `Drop` + `ZeroizeOnDrop` will be sufficient.

**Comment by newpavlov** [comment]
Also it's probably worth to wait for the next breaking release of `digest` which will bump MSRV to 1.65 and add `zeroize` support during migration to it.

**Comment by kayabaNerve** [comment]
This does not return an invalid state. It's effectively a reset with a guarantee the prior state is gone. Please read the implementation.

Also, I am unable to get CI happy 0_o

**Comment by newpavlov** [comment]
You literally set zeros into hasher states, which may cause security issues if such state is to be accidentally used by users. For proper reset you should use the initialization constants. You SHOULD NOT implement the `Zeroize` trait for types which may contain and rely on any kind of inner invariant. It's misuse of the trait.

**Comment by kayabaNerve** [comment]
... and then immediately after setting zeroes, set the initialization constants.

**Comment by kayabaNerve** [comment]
If you'd rather it directly set the initialization constants, that can be done. It'd just require re-implementing the zeroize's crate internal mechanisms to avoid being optimized out by the compiler. That would duplicate code (several times), notably increase maintenance effort, and be at greater risk of failure IMO.

**Comment by newpavlov** [comment]
Yes, you are right, but it's not "zeroization" anymore, isn't it? The point of the `Zeroize` trait is to perform "volatile" erasure of secrets and it's usually used to erase "primitive" values. Erasure of "complex" types is done using `Drop` impls which use `Zeroize` under the hood.

I will repeat myself: you should not implement the `Zeroize` trait for the hasher types. See the other algorithm crates which support `zeroize`, e.g. the block cipher crates.

**Comment by kayabaNerve** [comment]
It's not zeroization, yet the intent is to erase the prior state. This does perform exactly that erasure. I'm unsure of anyone who performs zeroization and then reads the memory, always expecting zeroes (no matter the type) like it's a `calloc` call.

As for whether or not this should *solely* implement `ZeroizeOnDrop`... Sure. I'm completely fine with that and can edit this PR to do it that way, per your preference :+1: I only did it this was initially as it leaves the user with greater choice over when they want to zeroize.

**Comment by tarcieri** [comment]
Generally `Zeroize` is useful for the low-level types that store state like integers and slices, and higher-level constructions benefit from `ZeroizeOnDrop` as there's no need to explicitly call `zeroize` (prevents that from being forgotten) and doing it on drop instead of manually prevents a use-after-zeroize condition.

With `ZeroizeOnDrop`, you can just enable the `zeroize` feature and you get zeroization for free with no additional changes.

**Comment by newpavlov** [comment]
The minimal versions CI failures are not relevant to this PR and should be fixed by https://github.com/RustCrypto/actions/pull/34

**Comment by kayabaNerve** [comment]
Happy to hear. I just pushed a commit which solely adds ZeroizeOnDrop for SHA 1..= 2 and Blake2 :+1:

File: sha3/Cargo.toml [hunk 1]
```diff
@@ -26,7 +26,7 @@ hex-literal = "0.2.2"
 
 [features]
 default = ["std"]
-std = ["digest/std"]
+std = ["digest/std", "zeroize?/std"]
```
Comment (newpavlov) [inline sha3/Cargo.toml]: Remove this line. It causes MSRV breakage.

File: blake2/Cargo.toml [hunk 2]
```diff
@@ -13,6 +13,7 @@ categories = ["cryptography", "no-std"]
 
 [dependencies]
 digest = { version = "0.10.7", features = ["mac"] }
+zeroize_crate = { package = "zeroize", version = "1", default-features = false, optional = true }
```
Comment (newpavlov) [inline blake2/Cargo.toml]: I think we can simply use `zeroize = { version = "1", default-features = false, optional = true }`?

Comment (kayabaNerve) [reply blake2/Cargo.toml]: I'm assuming you mean by removing the zeroize feature, which does seem to not be needed in this use-case. Sorry, I'm unfamiliar with writing tomls with such old support :sweat_smile:

**Comment by kayabaNerve** [comment]
Pushed what I believe to be resolutions :+1: Thanks for the tips :)

**Comment by kayabaNerve** [comment]
Is there anything I can do to push this forward?

**Comment by newpavlov** [comment]
Closing in favor of #545.

**Pull request closed without merging**